File: CTCI/chapter01/problem0101/Problem0101.py
# [1.1] Implement an algorithm to determine if a string has all unique characters.

def allUniqueChars(theString):
    # Optimization
    # if # chars > 256 then there must be duplicates
    if (len(theString) > 256):
        return 0;
    # The meat
    charCounts = [0] * 256;
    for char in theString:
        if (charCounts[ord(char)]):
            return 0;
        charCounts[ord(char)] = 1;
    return 1;

theString = "thisisastringwithrepeatingchars";
print "~Problem0101~";
print theString;
print allUniqueChars(theString);
theString = "asdfghjkl";
print theString;
print allUniqueChars(theString);

File: CTCI/chapter01/problem0101/Problem0101.cpp
#include <iostream>
#include <string>
using namespace std;

// [1.1] Implement an algorithm to determine if a string has all unique characters.

class Problem0101 {
public:
    bool allUniqueChars(string str);
};

bool Problem0101::allUniqueChars(string str) {
    // Optimization
    // If # chars > 256 there must be duplicates
    if (str.length() > 256) {
        return false;
    }
    // The meat
    bool charCount[256] = {false};
    for (int i=0; i<str.length(); i++) {
        int charVal = str.at(i);
        if (charCount[charVal]) {
            return false;
        }
        charCount[charVal] = true;
    }
    return true;
}

int main() {
    Problem0101 problem0101;
    cout << "~Problem0101~" << endl;
    string str = "thisisastringwithrepeatchars";
    cout << str << endl << problem0101.allUniqueChars(str) << endl;
    string str2 = "asdfghjkl";
    cout << str2 << endl << problem0101.allUniqueChars(str2) << endl;
    return 0;
}
